spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/discovery/DiscoveryClientRouteDefinitionLocator.java

```diff
@@ -159,6 +156,30 @@ private DiscoveryClientRouteDefinitionLocator(String discoveryClientName,
 				});
 	}
 
+	protected RouteDefinition buildRouteDefinition(Expression urlExpr, ServiceInstance serviceInstance) {
+		String serviceId = serviceInstance.getServiceId();
+		RouteDefinition routeDefinition = new RouteDefinition();
+		routeDefinition.setId(this.routeIdPrefix + serviceId);
+		String uri = urlExpr.getValue(this.evalCtxt, serviceInstance, String.class);
+		routeDefinition.setUri(URI.create(uri));
+		// add instance metadata
+		Map<String, Object> metadata = new HashMap<>(serviceInstance.getMetadata());
+		for (Map.Entry<String, Object> entry : metadata.entrySet()) {
```

Comment:
Let's avoid any conversions here. I have some ideas for how to do conversions, but it will be more involved.

Reply:
fd1763d5b9f2b2f744d87f4ed0a41d7c7678a896 added support for those keys as strings.